File: MCPForUnity/Editor/Tools/ReloadDomain.cs
using System;
using MCPForUnity.Editor.Helpers;
using Newtonsoft.Json.Linq;
using UnityEditor;

namespace MCPForUnity.Editor.Tools
{
    /// <summary>
    /// Handles domain reload operations to refresh Unity's script assemblies.
    /// This is essential after creating or modifying scripts to make new types available.
    /// </summary>
    [McpForUnityTool("reload_domain")]
    public static class ReloadDomain
    {
        /// <summary>
        /// Main handler for domain reload command.
        /// Triggers Unity to reload all script assemblies, which is necessary after
        /// script changes before new components can be used.
        /// </summary>
        public static object HandleCommand(JObject @params)
        {
            try
            {
                McpLog.Info("[ReloadDomain] Requesting domain reload");
                EditorUtility.RequestScriptReload();
                return Response.Success("Domain reload requested. Unity will recompile scripts and refresh assemblies.");
            }
            catch (Exception ex)
            {
                McpLog.Error($"[ReloadDomain] Error requesting domain reload: {ex.Message}");
                return Response.Error($"Failed to request domain reload: {ex.Message}");
            }
        }
    }
}

File: Server/tools/reload_domain.py
from fastmcp import Context
from models import MCPResponse
from registry import mcp_for_unity_tool
from unity_connection import async_send_command_with_retry


@mcp_for_unity_tool(
    description="Trigger a Unity domain reload to recompile scripts and refresh assemblies. Essential after creating or modifying scripts before new components can be used."
)
async def reload_domain(ctx: Context) -> MCPResponse:
    """
    Request Unity to reload its domain (script assemblies).
    This is necessary after:

    - Creating new C# scripts
    - Modifying existing scripts
    - Before attempting to add new components to GameObjects

    Returns immediately after triggering the reload request.
    Unity will handle the actual recompilation asynchronously.
    """
    await ctx.info("Requesting Unity domain reload")
    result = await async_send_command_with_retry("reload_domain", {})
    return MCPResponse(**result) if isinstance(result, dict) else result
